### ISSUE TITLE
SymbolicExpression.Parse freeze for "29^11^7"

### ISSUE THREAD

**Issue by akolevmm** [mathnet/mathnet-symbolics#64]
Using C# nuget package MathNet.Symbolics, ver 0.20.0

Parsing this expression "29^11^7" leads to freeze.
Tried with 
Infix.ParseOrUndefined("29^11^7");
and
SymbolicExpression.Parse("29^11^7")

When tried to put parenthesis it works fine
"(29^11)^7"

**Comment by FoggyFinder** [comment]
I think it's just trying to calculate:

"29^(11^7)" 

that's very big number. So, put a parentheses is a good idea.

**Comment by Happypig375** [comment]
Yes. Exponentiation is right-associative, which means that it is evaluated right-to-left.

**Comment by Happypig375** [comment]
I think large number detection can be done to avoid hangs.

**Comment by akolevmm** [comment]
Yes, it sounds reasonable.
The problem in my case with this expression is that it is entered by user and I don't have control over what he can enter. 
Is there a way to handle this?

**Comment by Happypig375** [comment]
Evaluate asynchronously and provide a cancel button?

**Comment by akolevmm** [comment]
> Evaluate asynchronously, show an activity indicator and provide a cancel button?

This is not possible, because my code is in web service invoked by other applications.

**Comment by Happypig375** [comment]
Setting a default timeout so that each request will not take too much time?

**Comment by akolevmm** [comment]
This is not an option too, because I have to return meaningful result in short time.
I have to somehow handle expressions like this without affecting the whole functionality.
For now my workaround is to run it in separate thread and wait for specified number of milliseconds, but I prefer to have better way to do it.

**Comment by cdrnet** [comment]
I guess we could have a global Control instance like in Numerics to control how a few operations like power behave on very large input (or expected very large output or expensive operation). Ugly, but usually good enough.

Any proposals on what good rules would be for power?

**Comment by MovGP0** [comment]
Instead of a default timeout, there should be 

- no eager execution. the user needs to explicitly apply an `Expression.simplify` function
- calculations should be done with `CancellationToken` support 

Using a `CancellationToken` allows for scenarios with cancel buttons, web requests, as well as timeouts (via the `CancellationTokenSource`).
